### PR TITLE
rewrote net

### PR DESCRIPTION
**Pull request by yuanqing-wang**

### CODE REVIEW

File: pinot/net.py [hunk 1]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
     learning, puts into a predicted distribution and calculates the
     corresponding divergence.
 
+
+    Attributes
+    ----------
+    representation: a `pinot.representation` module
+        the model that translates graphs to latent representations
+    output_regression: a `torch.nn.Moudle` or None,
```
Comment (karalets) [inline pinot/net.py:26]: typo: nn.Module

File: pinot/net.py [hunk 2]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+        if None, this will be set as a simple `Linear` layer that inputs
+        the latent dimension and output the number of parameters for
+        `self.distribution_class`
+    distribution_class: a `torch.distribution` object,
```
Comment (karalets) [inline pinot/net.py:30]: Do we need both this and the noise model? distribution class plays the role of the noise model here. What is your reasoning for this redundancy?

File: pinot/net.py [hunk 3]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+
+
     """
 
     def __init__(
         self,
         representation,
-        parameterization,
+        output_regression=None,
         distribution_class=torch.distributions.normal.Normal,
```
Comment (karalets) [inline pinot/net.py:47]: same thing as above... I would imagine all of this can be absorbed into noise model.

File: pinot/net.py [hunk 4]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+        
         super(Net, self).__init__()
         self.representation = representation
-        self.parametrization = parameterization
         self.distribution_class = distribution_class
-        self.param_transform = param_transform
-        self.expectation_fn = expectation_fn
+
+        # grab the last dimension of `representation`
+        regression_in_features = [
```
Comment (karalets) [inline pinot/net.py:56]: I'd call it 'representation_hidden' or so but we can nitpick names later.

File: pinot/net.py [hunk 5]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+                layer for layer in list(self.representation.modules())\
+                        if hasattr(layer, 'out_features')][-1].out_features
+
+        # TODO:
+        # make this less ugly
+        regression_out_features = len(distribution_class.arg_constraints)
```
Comment (karalets) [inline pinot/net.py:62]: yeah this is uglier than necessary, a noise model as I said earlier specifies how to handle this implicitly.
currently you have many different locations where you accidentally have to get things right although they all have the same cause: the choice of noise model.

File: pinot/net.py [hunk 6]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+
+        if output_regression is None:
+            # make the output regression as simple as a linear one
+            # if nothing is specified
+            self._output_regression = torch.nn.ModuleList(
+                    [
+                        torch.nn.Linear(regression_in_features, 1)\
```
Comment (karalets) [inline pinot/net.py:69]: why (regression_in_features, 1) ?

File: pinot/net.py [hunk 7]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+                                for _ in range(regression_out_features)
+                    ])
+
+            def output_regression(theta):
```
Comment (karalets) [inline pinot/net.py:73]: does it need that? imho if you build an nn.sequential with a modulelist object inside, i.e.

regressor_net = nn.Sequential(self._output_regression) 

you should be able to just call .forward() on that later and get back your list without needing a for loop through these modules.

File: pinot/net.py [hunk 8]
```diff
@@ -18,22 +18,78 @@ class Net(torch.nn.Module):
...
+                return [f(theta) for f in self._output_regression]
+
+        self.output_regression = output_regression
+
+        # TODO:
+        # make this less ugly
+        if isinstance(noise_model, str):
```
Comment (karalets) [inline pinot/net.py:80]: yeah this to me again makes no sense or I do not understand your design choices. Why do we still need all this stuff here?

All of these choices are model choices and should live in the model, i.e. where you use distribution.

Why is this here?

File: pinot/net.py [hunk 9]
```diff
@@ -56,28 +112,21 @@ def condition(self, g):
...
+        # put theta through noise model
+        theta = self.noise_model(theta)
+
         # parameterize the distribution according to the parameters
         # and the distribution class
-
-        distribution = self.distribution_class(
-            *self.param_transform(*torch.unbind(theta, dim=-1))
-        )
+        distribution = self.distribution_class(*theta)
```
Comment (karalets) [inline pinot/net.py:120]: why not use the noise model flag here and drop this contrived distribution_class object?
It would simplify so many things above...

Is there a strong reason to keep this?

File: pinot/net.py [hunk 10]
```diff
@@ -56,28 +112,21 @@ def condition(self, g):
...
-        """ Parameterize a distribution and sample from it.
-        """
-        distribution = self.condition(g)
-
-        if n_samples == 0:
-            return distribution.sample()
-
-        return distribution.sample_n(n_samples)
-
     def expectation(self, g):
```
Comment (karalets) [inline pinot/net.py:124]: Why do we need this expectation thing?

File: pinot/net.py (hunk 2 above)
Comment (karalets) [reply pinot/net.py:30]: And to clarify: I would prefer to only pass noise_model and subject to the choice of noise model construct the right likelihood later on in one of the forward-type functions.

Why have all this duplication?

File: pinot/net.py (hunk 6 above)
Comment (karalets) [reply pinot/net.py:69]: To be precise: my question is about the 1.
Should this not be some output_dim ? I.e. the dimensionality of our measurements?

File: pinot/net.py (hunk 2 above)
Comment (yuanqing-wang) [reply pinot/net.py:30]: I was trying to allow manipulation of sigma.

Comment (karalets) [reply pinot/net.py:30]: what do you mean?

File: pinot/net.py (hunk 7 above)
Comment (yuanqing-wang) [reply pinot/net.py:73]: here we're using `torch.nn.ModuleList` and the `forward` function of it raises NotImplemented Error

File: pinot/net.py (hunk 9 above)
Comment (yuanqing-wang) [reply pinot/net.py:120]: well it's not necessarily Normal

File: pinot/net.py (hunk 10 above)
Comment (yuanqing-wang) [reply pinot/net.py:124]: I thought it would be interesting to keep the functionality to enable training on mean estimate.

File: pinot/net.py (hunk 7 above)
Comment (karalets) [reply pinot/net.py:73]: did you try first wrapping it into a sequential?

File: pinot/net.py (hunk 9 above)
Comment (karalets) [reply pinot/net.py:120]: the noise model 'homoschedastic_normal' implies normal with fixed global variance.
The distribution and its parametrization are both properties of noise_model.
If you want another distribution, you are picking another likelihood which is another noise model. 

I would put all this into noise model or likelihood model or however you want to call it.

Passing the distribution separately makes no sense, these two objects are bound together.

Comment (karalets) [reply pinot/net.py:120]: I would absorb everything into noise model like this, as I also suggested a long time ago in the modeling issue:
```
def forward/condition/whatever():
    if noise_model=='normal_something_schedastic'
        mu = ...
        log_sigma = ....
        sigma = torch.exp(sigma)
        distribution = Normal(mu,sigma)
    if noise_model == 'categorical':
        alpha = ... #i.e. output from a regressor
        dist_noise = Categorical(alpha)
```

Comment (karalets) [reply pinot/net.py:120]: A HUGE advantage of this is that we keep the likelihood very interpretable code-wise ad do not mix it with representation stuff.

I think we will have another small refactoring step here later on when we try things that are not simple linear decoders/regressors, but we can cross that bridge when we get there.

File: pinot/net.py (hunk 10 above)
Comment (karalets) [reply pinot/net.py:124]: well you can always do loss.mean() after you return the loss, no need to write an extra function in here for that, no?

Comment (karalets) [reply pinot/net.py:124]: maybe I am missing the point here...

File: pinot/net.py (hunk 9 above)
Comment (karalets) [reply pinot/net.py:120]: @yuanqing-wang  so I have this alternative pitch if you prefer it.

You can create a meta-object which ties together parametrization and distribution and passes them in here instead of a flag if you prefer.

i.e. outside of the model you can create that thing as
```
class HomoSchedasticNormal():
    def __init()__:
    .....properties and parametrizations
 ```

File: pinot/net.py (hunk 7 above)
Comment (karalets) [reply pinot/net.py:73]: i.e. ```regressor =  nn.Sequential(module_list_you_created)``` should work with .forward()

Comment (karalets) [reply pinot/net.py:73]: it might need an identity layer pre-prended or so, let me know if it does not work out of the box.

Comment (karalets) [reply pinot/net.py:73]: ok I think this is resolved, you indeed need that for loop

File: pinot/net.py [hunk 11]
```diff
@@ -109,24 +87,32 @@ def forward(self, g):
...
-        # parameterize the distribution according to the parameters
-        # and the distribution class
-        distribution = self.distribution_class(*theta)
+        elif self.noise_model == 'normal-homoschedastic':
+            mu, _ = self.forward(g)
 
-        return distribution
+            # initialize a `LOG_SIMGA` if there isn't one
+            if not hasattr(self, 'LOG_SIGMA'):
+                self.LOG_SIGMA = torch.tenosr(0.0)
```
Comment (karalets) [inline pinot/net.py:102]: typo, do: ```torch.tensor```

And here now you can also bring in the measurement dimension, since you will have a log_sigma per dimension instead of a scalar.

File: pinot/net.py [hunk 12]
```diff
@@ -109,24 +87,32 @@ def forward(self, g):
...
-        return distribution.mean()
+            distribution = torch.distributions.normal.Normal(
+                    loc=mu,
+                    scale=torch.exp(self.LOG_SIGMA))
+
+        elif self.noise_model == 'normal-homoschedastic-fixed':
+            mu, _ = self.forward(g)
+            distribution = torch.distributions.normal.Normal(
+                    loc=mu, 
+                    scale=torch.tensor(1.0))
```
Comment (karalets) [inline pinot/net.py:113]: similar to above, instead of 1.0 I'd use sth like ''' torch.ones(1,measurement_dim)"'' or whatever it is called

Comment (karalets) [reply pinot/net.py:113]: also ```torch.ones(measurement_dim)``` will broadcast automagically in pytorch, it is just somewhat inconsistent with the other shapes

**Comment by karalets** [comment]
Let me add one more comment:
OI think it wopuld be good if you added a global variable ```self.noise_model_args``` which can be ```None``` typically, but could contain arguments that are relevant for arbitrary settings we may want to have.

I.e. when you have fixed variance per default we set it to 1.0.
There may e cases where we may want to set it to some value based on normalization etc.
Int hat case you could have ```noise_model_args.std = value``` and then use that inside the noise model if this is passed.

For more complex noise models one may want to add later it gives flexibility to add more details, as these args can be whatever they want to be and just inform the noise model.

What do you think?

**Comment by yuanqing-wang** [comment]
how about, leave the possibility open by allowing users to specify a dict about how they would like to handle distribution parameters and distribution class, namely

```
        else:
            assert isinstance(
                    self.noise_model,
                    dict)

            distribution = self.noise_model['distribution'](
                    self.noise_model['kwargs'])
```

**Comment by yuanqing-wang** [comment]
@karalets if you think the current implementation is ok I'll go ahead and rewrite many of the tests

**Comment by karalets** [comment]
> how about, leave the possibility open by allowing users to specify a dict about how they would like to handle distribution parameters and distribution class, namely
> 
> ```
>         else:
>             assert isinstance(
>                     self.noise_model,
>                     dict)
> 
>             distribution = self.noise_model['distribution'](
>                     self.noise_model['kwargs'])
> ```

hmm i don't love this exact design, we can leave it for later

**Comment by karalets** [comment]
> @karalets if you think the current implementation is ok I'll go ahead and rewrite many of the tests

yeah try to test this now, please

File: pinot/net.py [hunk 13]
```diff
@@ -99,7 +99,7 @@ def condition(self, g):
 
             # initialize a `LOG_SIMGA` if there isn't one
             if not hasattr(self, 'LOG_SIGMA'):
-                self.LOG_SIGMA = torch.tenosr(0.0)
+                self.LOG_SIGMA = torch.zeros((1, measurement_dimension))
```
Comment (karalets) [inline pinot/net.py:102]: does ```measurement_dimension``` not need a self?

File: pinot/net.py [hunk 14]
```diff
@@ -110,7 +110,17 @@ def condition(self, g):
             mu, _ = self.forward(g)
             distribution = torch.distributions.normal.Normal(
                     loc=mu, 
-                    scale=torch.tensor(1.0))
+                    scale=torch.ones((1, measurement_dimension)))
```
Comment (karalets) [inline pinot/net.py:113]: same as above

Comment (yuanqing-wang) [reply pinot/net.py]: ops sorry

fixed here 
027d5c7
